libservo: Add `WebViewDelegate` and `ServoDelegate` and port `winit_minimal`

delan opened the pull request:
This change adds the second major part of the new API: delegates which
have methods called by the Servo loop. When a delegate is set on a
`WebView` or on `Servo` itself, the event loop will call into
appropriate delegate methods. Applications can implement the delegate on
their own structs to add special behavior per-`WebView` or for all
`WebView`s.

In addition, each delegate has a default implementation, which
automatically exposes "reasonable" behavior such as by-default allowing
navigation.

There's a lot more work to do here, such as refining the delegate
methods so that they all have nice interfaces, particulary with regard
to delegate methods that need an asynchronous response. This will be
handed gradually as we keep working on the API.

Co-authored-by: Martin Robinson <mrobinson@igalia.com>
Co-authored-by: Mukilan Thiyagarajan <mukilan@igalia.com>
Signed-off-by: Delan Azabani <dazabani@igalia.com>

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

wusyong commented on components/servo/webview.rs:
```diff
@@ -7,21 +7,149 @@ use std::ffi::c_void;
+    fn notify_devtools_server_started(_: Result<u16, ()>, _: String);
+}
+
+pub trait WeDidntGetTimeForThisDelegateButItWasFormerlyWindowMethods {
+    fn get_coordinates(&self) -> compositing::windowing::EmbedderCoordinates;
+    fn notify_animation_state_changed(&self, state: compositing::windowing::AnimationState);
+}
```
Here's the discussion that animation state change can move to servo delegate. I would expect both stateful and immediate mode would like to keep polling when the state change to animating.

`EmbedderCoordinates`, however, is more tricky. I hope we can remove it and find out what exact values they want from each call.

mrobinson commented on components/servo/webview.rs:
```diff
@@ -7,21 +7,149 @@ use std::ffi::c_void;
+    // fn load_complete(); // see notify_loading_state_changed
+    /// The loading state of the page has changed.
+    fn notify_loading_state_changed(_: WebView, _: ()); // TODO: make an enum for this
+    /// EmbeddderMsg::WebResourceRequested had no doc comment
+    // unsure what uses this and unsure of its semantics
+    // fn web_resource_requested(_: WebView, _: WebResourceRequest, _: IpcSender<WebResourceResponseMsg>);
```
It doesn't look like servoshell is every responding to this new message so we are confused about how this works.

jdm replied:
The network code treats a disconnected channel as the embedder ignoring the messages.

wusyong commented:
Really happy to see it's out of the draft, but it seems to become thousands of LOC.
This is similar to #34371, and I believe the feedback is against a sudden big chunk of code. That PR is even a port that won't affect the existing codebase.
I wonder if it's possible to split some of the changes to other PRs so that it's easier for us to review. Perhaps some variable name and trait methods don't need to be in this one.
Again, it's very cool to see more ergonomic webview APIs, and let's see how they could improve the user experience!

mrobinson commented:
> Really happy to see it's out of the draft, but it seems to become thousands of LOC. This is similar to #34371, and I believe the feedback is against a sudden big chunk of code. 

Yeah, this is really unfortunate, but a lot of this is really hard to split up. We could probably land the changes to `servoshell` separately, but the issue is that switching `servoshell` to use delegates necessitated reworking a lot of the code. Right now `servoshell` is written with the idea of mutable objects passed around to handlers in  a top down way. Using a delegate required inverting the ownership which is one reason why there are so many lines of code changed here.

mrobinson commented:
Just a note that we feel more confident about a big change like this, because it was done by the three of us working together pair programming, catching bugs as we went. That said, I can try to split out the `servoshell` changes into a separate PR.

delan committed:
libservo: Add WebViewDelegate and ServoDelegate and port `winit_minimal`

This change adds the second major part of the new API: delegates which
have methods called by the Servo loop. When a delegate is set on a
`WebView` or on `Servo` itself, the event loop will call into
appropriate delegate methods. Applications can implement the delegate on
their own structs to add special behavior per-`WebView` or for all
`WebView`s.

In addition, each delegate has a default implementation, which
automatically exposes "reasonable" behavior such as by-default allowing
navigation.

There's a lot more work to do here, such as refining the delegate
methods so that they all have nice interfaces, particulary with regard
to delegate methods that need an asynchronous response. This will be
handed gradually as we keep working on the API.

Co-authored-by: Martin Robinson <mrobinson@igalia.com>
Co-authored-by: Mukilan Thiyagarajan <mukilan@igalia.com>
Signed-off-by: Delan Azabani <dazabani@igalia.com>
Signed-off-by: Martin Robinson <mrobinson@igalia.com>

mrobinson commented:
@wusyong If you like to take another look at this, it's much smaller now without the servoshell changes. Most of the code modifications here are the addition of the boilerplate delegate traits.

wusyong commented on components/servo/lib.rs:
```diff
@@ -527,14 +540,24 @@ impl Servo {
+        self.delegate.clone()
+    }
+
+    pub fn set_delegate(&mut self, delegate: Rc<dyn ServoDelegate>) {
+        self.delegate = delegate;
+    }
```
Do we need to set the delegate dynamically? I feel like it can just be a parameter for `new` method.

mrobinson replied:
Yes, this is a pretty central part of the API. The idea is that the `WebView` should work even with no delegate set. The idea being that the only thing you need to do to get a working application is to create a `WebView` and add it to your interface.

mrobinson replied:
This is essentially the same pattern that WebKit uses FWIW: https://stackoverflow.com/questions/57144195/uidelegate-method-create-webview-with-configuration-not-called

wusyong commented on components/servo/webview.rs:
```diff
@@ -49,103 +73,217 @@ impl WebView {
+        self.inner().delegate.clone()
+    }
+
+    pub fn set_delegate(&self, delegate: Rc<dyn WebViewDelegate>) {
+        self.inner_mut().delegate = delegate;
     }
```
Ditto

wusyong commented on components/servo/webview.rs:
```diff
@@ -2,34 +2,58 @@
 
 #[derive(Clone)]
-pub struct WebView(Rc<WebViewInner>);
+pub struct WebView(Rc<RefCell<WebViewInner>>);
```
This doesn't need to change now, but I think WebView handle might need to cross the threads if the embedders have this demand. It will probably be `Arc<Mutex<T>>` in the future.

mrobinson replied:
Yes, we have discussed this. It would likely require a lot more locking. For now, embedders can do their own cross-thread communication to ensure that Servo stay on a single thread though.

wusyong commented on components/servo/lib.rs:
```diff
@@ -961,8 +986,36 @@ impl Servo {
+    ///   - Runs delebgate methods on all `WebView`s and `Servo` itself
+    ///   - Maybe update the rendered compositor output, but *without* swapping buffers.
+    ///
+    pub fn spin_event_loop(&self) -> bool {
```
Yes it should explain what the return value is.

mrobinson replied:
Okay. I've added some more documentation here, clarifying that false means that you should no longer spin the Servo event loop and that shutdown is complete.

wusyong commented on components/servo/lib.rs:
```diff
@@ -961,8 +986,36 @@ impl Servo {
+
+    /// Spin the Servo event loop, which:
+    ///
+    ///   - Performs updates in the compositor, such as queued pinch zoom events
+    ///   - Runs delebgate methods on all `WebView`s and `Servo` itself
+    ///   - Maybe update the rendered compositor output, but *without* swapping buffers.
```
What I see is compositor "receive messages first -> call delegate methods -> perform update." I think it should either be more transparent about what it exactly calls or update the compositor's method names to reflect what we want to explain here.

mrobinson replied:
Agreed. We actually plan to update the compositor names very soon to make it a lot clearer what they do.

wusyong reviewed (changes requested):
Overall LGTM too. I just have a few questions and want to improve some document comments.

mrobinson commented on components/servo/webview.rs:
```diff
@@ -2,34 +2,58 @@
+    }
+}
+
+impl Hash for WebView {
```
Yep, this is so `WebView` can be stored in `HashSet`.

mrobinson commented:
@wusyong I think I've addressed your comments, please take another look.
